File: gc_test/object_graph/object_graph.c
// Test case that builds an object graph similar to the one used in tmux.  The
// graph consists of sessions and windows; each session can have several
// windows, and each window can be associated with several sessions.  A session
// with no windows or a window with no sessions will be implicitly destroyed.

#include <stdlib.h>
#include <stdio.h>
#include <sys/queue.h>

// If set, also define globals lists of all windows and all sessions.
//#define GLOBAL_LISTS

struct session;
struct window;

struct link {
    struct session* session;
    struct window* window;
    // An entry in the session's list of links.
    TAILQ_ENTRY(link) session_entry;
    // An entry in the window's list of links.
    TAILQ_ENTRY(link) window_entry;
};
TAILQ_HEAD(links, link);

struct session {
    int id;
    TAILQ_ENTRY(session) global_entry;
    struct links links;
};
#ifdef GLOBAL_LISTS
TAILQ_HEAD(sessions, session);
static struct sessions global_sessions = TAILQ_HEAD_INITIALIZER(global_sessions);
#endif // GLOBAL_LISTS
static int next_session_id = 0;

struct window {
    int id;
    TAILQ_ENTRY(window) global_entry;
    struct links links;
};
#ifdef GLOBAL_LISTS
TAILQ_HEAD(windows, window);
static struct windows global_windows = TAILQ_HEAD_INITIALIZER(global_windows);
#endif // GLOBAL_LISTS
static int next_window_id = 0;



struct session* session_new();
void session_delete(struct session* sess);
void session_render(struct session* sess);

struct window* window_new(struct session* sess);
void window_delete(struct window* win);

void window_link(struct window* win, struct session* sess);
void window_unlink(struct window* win, struct session* sess);
#define CLEANUP_SESSION 1
#define CLEANUP_WINDOW 2
void link_delete(struct link* l, int cleanup_flags);


struct session* session_new() {
    struct session* sess = calloc(1, sizeof(struct session));
    sess->id = ++next_session_id;
    printf("new session %d\n", sess->id);
    TAILQ_INIT(&sess->links);
#ifdef GLOBAL_LISTS
    TAILQ_INSERT_TAIL(&global_sessions, sess, global_entry);
#endif

    // Create an initial window for the session.
    (void)window_new(sess);

    return sess;
}

void session_delete(struct session* sess) {
    printf("delete session %d\n", sess->id);
    // Unlink all windows
    struct link* l = TAILQ_FIRST(&sess->links);
    while (l != NULL) {
        struct link* next = TAILQ_NEXT(l, session_entry);
        link_delete(l, CLEANUP_WINDOW);
        l = next;
    }

    // Destroy the session
#ifdef GLOBAL_LISTS
    TAILQ_REMOVE(&global_sessions, sess, global_entry);
#endif
    printf("deleted session %d\n", sess->id);
    free(sess);
}

void session_render(struct session* sess) {
    printf("session %d windows: ", sess->id);
    struct link* l;
    int first = 1;
    TAILQ_FOREACH(l, &sess->links, session_entry) {
        if (first) {
            first = 0;
            printf("%d", l->window->id);
        } else {
            printf(", %d", l->window->id);
        }
    }
    if (first) {
        // No links were found - should be impossible
        printf("(none)\n");
    } else {
        printf("\n");
    }
}


struct window* window_new(struct session* sess) {
    struct window* win = calloc(1, sizeof(struct window));
    win->id = ++next_window_id;
    printf("new window %d\n", win->id);
    TAILQ_INIT(&win->links);
#ifdef GLOBAL_LISTS
    TAILQ_INSERT_TAIL(&global_windows, win, global_entry);
#endif

    window_link(win, sess);

    return win;
}

void window_delete(struct window* win) {
    printf("delete window %d\n", win->id);

    // Unlink all sessions
    struct link* l = TAILQ_FIRST(&win->links);
    while (l != NULL) {
        struct link* next = TAILQ_NEXT(l, window_entry);
        link_delete(l, CLEANUP_SESSION);
        l = next;
    }

    // Destroy the window
#ifdef GLOBAL_LISTS
    TAILQ_REMOVE(&global_windows, win, global_entry);
#endif
    printf("deleted window %d\n", win->id);
    free(win);
}

void window_link(struct window* win, struct session* sess) {
    printf("link session %d, window %d\n", sess->id, win->id);
    struct link* link = calloc(1, sizeof(struct link));
    link->session = sess;
    link->window = win;
    TAILQ_INSERT_TAIL(&sess->links, link, session_entry);
    TAILQ_INSERT_TAIL(&win->links, link, window_entry);
}

void window_unlink(struct window* win, struct session* sess) {
    struct link* l;
    TAILQ_FOREACH(l, &win->links, window_entry) {
        if (l->session == sess) {
            break;
        }
    }
    if (l != NULL) {
        link_delete(l, CLEANUP_SESSION | CLEANUP_WINDOW);
    }
}

void link_delete(struct link* l, int cleanup_flags) {
    printf("unlink session %d, window %d\n", l->session->id, l->window->id);

    if (cleanup_flags & CLEANUP_SESSION) {
        TAILQ_REMOVE(&l->session->links, l, session_entry);
        if (TAILQ_EMPTY(&l->session->links)) {
            session_delete(l->session);
        }
    }

    if (cleanup_flags & CLEANUP_WINDOW) {
        TAILQ_REMOVE(&l->window->links, l, window_entry);
        if (TAILQ_EMPTY(&l->window->links)) {
            window_delete(l->window);
        }
    }

    free(l);
}


int main() {
    struct session* sess1 = session_new();
    struct session* sess2 = session_new();

    printf("\ndelete a window explicitly, which removes it from its sessions\n");
    struct window* win3 = window_new(sess1);
    window_link(win3, sess2);

    session_render(sess1);
    session_render(sess2);

    window_delete(win3);

    session_render(sess1);
    session_render(sess2);

    printf("\ndelete a window implicitly by removing it from all sessions\n");
    struct window* win4 = window_new(sess1);
    window_unlink(win4, sess1);

    session_render(sess1);
    session_render(sess2);

    printf("\ndelete a session implicitly by removing all of its windows\n");
    struct session* sess3 = session_new();
    struct window* win5 = TAILQ_FIRST(&sess3->links)->window;
    window_unlink(win5, sess3);

    printf("\ndelete a session implicitly by deleting all of its windows\n");
    struct session* sess4 = session_new();
    struct window* win6 = TAILQ_FIRST(&sess4->links)->window;
    window_delete(win6);

    printf("\ndelete sessions, which removes and deletes all of their windows\n");
    struct window* win7 = window_new(sess1);
    window_link(win7, sess2);

    session_render(sess1);
    session_render(sess2);

    session_delete(sess2);
    session_delete(sess1);
}

File: gc_test/object_graph/object_graph.rs
#![allow(dead_code, mutable_transmutes, non_camel_case_types, non_snake_case, non_upper_case_globals, unused_assignments, unused_mut)]

use std::cell::Cell;
use std::sync::atomic::{AtomicI32, Ordering};
use gc_lib::cell2::Cell2;
use gc_lib::drc_gc::{NullableDrc, NullableSubDrc, BreakCycles};

#[derive(Clone)]
#[repr(C)]
pub struct session {
    pub id: Cell<libc::c_int>,
    pub links: links,
}
#[derive(Clone)]
#[repr(C)]
pub struct links {
    pub tqh_first: Cell2<NullableDrc<link>>,
    pub tqh_last: Cell2<NullableSubDrc<Cell2<NullableDrc<link>>>>,
}
#[derive(Clone)]
#[repr(C)]
pub struct link {
    pub session: Cell2<NullableDrc<session>>,
    pub window: Cell2<NullableDrc<window>>,
    pub session_entry: tailq_entry_link,
    pub window_entry: tailq_entry_link,
}
#[derive(Clone)]
#[repr(C)]
pub struct tailq_entry_link {
    pub tqe_next: Cell2<NullableDrc<link>>,
    pub tqe_prev: Cell2<NullableSubDrc<Cell2<NullableDrc<link>>>>,
}
#[derive(Clone)]
#[repr(C)]
pub struct window {
    pub id: Cell<libc::c_int>,
    pub links: links,
}
static next_session_id: AtomicI32 = AtomicI32::new(0 as libc::c_int);
static next_window_id: AtomicI32 = AtomicI32::new(0 as libc::c_int);

impl BreakCycles for session {
    fn break_cycles(&self) {
        self.links.break_cycles();
    }
}

impl BreakCycles for links {
    fn break_cycles(&self) {
        self.tqh_first.set(NullableDrc::null());
        self.tqh_last.set(NullableSubDrc::null());
    }
}

impl BreakCycles for link {
    fn break_cycles(&self) {
        self.session.set(NullableDrc::null());
        self.window.set(NullableDrc::null());
        self.session_entry.break_cycles();
        self.window_entry.break_cycles();
    }
}

impl BreakCycles for tailq_entry_link {
    fn break_cycles(&self) {
        self.tqe_next.set(NullableDrc::null());
        self.tqe_prev.set(NullableSubDrc::null());
    }
}

impl BreakCycles for window {
    fn break_cycles(&self) {
        self.links.break_cycles();
    }
}

#[no_mangle]
pub fn session_new() -> NullableDrc<session> {
    let sess: NullableDrc<session> = NullableDrc::new(session {
        id: Cell::new(0),
        links: links {
            tqh_first: Cell2::new(NullableDrc::null()),
            tqh_last: Cell2::new(NullableSubDrc::null()),
        },
    });
    next_session_id.fetch_add(1, Ordering::Relaxed);
    sess.id.set(next_session_id.load(Ordering::Relaxed));
    println!("new session {}", sess.id.get());
    sess.links.tqh_first.set(NullableDrc::null());
    sess.links.tqh_last.set(sess.project(|sess| &sess.links.tqh_first));
    window_new(sess);
    sess
}
#[no_mangle]
pub fn session_delete(mut sess: NullableDrc<session>) {
    println!("delete session {}", sess.id.get());
    let mut l: NullableDrc<link> = sess.links.tqh_first.get();
    while !l.is_null() {
        let mut next: NullableDrc<link> = l.session_entry.tqe_next.get();
        link_delete(l, 2 as libc::c_int);
        l = next;
    }
    println!("deleted session {}", sess.id.get());
    sess.drop_data();
}
#[no_mangle]
pub fn session_render(mut sess: NullableDrc<session>) {
    print!("session {} windows: ", sess.id.get());
    let mut l: NullableDrc<link> = NullableDrc::null();
    let mut first: libc::c_int = 1 as libc::c_int;
    l = sess.links.tqh_first.get();
    while !l.is_null() {
        if first != 0 {
            first = 0 as libc::c_int;
            print!("{}", l.window.get().id.get());
        } else {
            print!(", {}", l.window.get().id.get());
        }
        l = l.session_entry.tqe_next.get();
    }
    if first != 0 {
        println!("(none)");
    } else {
        println!();
    };
}
#[no_mangle]
pub fn window_new(mut sess: NullableDrc<session>) -> NullableDrc<window> {
    let mut win: NullableDrc<window> = NullableDrc::new(window {
        id: Cell::new(0),
        links: links {
            tqh_first: Cell2::new(NullableDrc::null()),
            tqh_last: Cell2::new(NullableSubDrc::null()),
        },
    });
    next_window_id.fetch_add(1, Ordering::Relaxed);
    win.id.set(next_window_id.load(Ordering::Relaxed));
    println!("new window {}", win.id.get());
    win.links.tqh_first.set(NullableDrc::null());
    win.links.tqh_last.set(win.project(|win| &win.links.tqh_first));
    window_link(win, sess);
    win
}
#[no_mangle]
pub fn window_delete(mut win: NullableDrc<window>) {
    println!("delete window {}", win.id.get());
    let mut l: NullableDrc<link> = win.links.tqh_first.get();
    while !l.is_null() {
        let mut next: NullableDrc<link> = l.window_entry.tqe_next.get();
        link_delete(l, 1 as libc::c_int);
        l = next;
    }
    println!("deleted window {}", win.id.get());
    win.drop_data();
}
#[no_mangle]
pub fn window_link(mut win: NullableDrc<window>, mut sess: NullableDrc<session>) {
    println!(
        "link session {}, window {}",
        sess.id.get(),
        win.id.get(),
    );
    let mut link: NullableDrc<link> = NullableDrc::new(link {
        session: Cell2::new(NullableDrc::null()),
        window: Cell2::new(NullableDrc::null()),
        session_entry: tailq_entry_link {
            tqe_next: Cell2::new(NullableDrc::null()),
            tqe_prev: Cell2::new(NullableSubDrc::null()),
        },
        window_entry: tailq_entry_link {
            tqe_next: Cell2::new(NullableDrc::null()),
            tqe_prev: Cell2::new(NullableSubDrc::null()),
        },
    });
    link.session.set(sess);
    link.window.set(win);
    link.session_entry.tqe_next.set(NullableDrc::null());
    link.session_entry.tqe_prev.set(sess.links.tqh_last.get());
    (*sess.links.tqh_last.get()).set(link);
    sess.links.tqh_last.set(link.project(|link| &link.session_entry.tqe_next));
    link.window_entry.tqe_next.set(NullableDrc::null());
    link.window_entry.tqe_prev.set(win.links.tqh_last.get());
    (*win.links.tqh_last.get()).set(link);
    win.links.tqh_last.set(link.project(|link| &link.window_entry.tqe_next));
}
#[no_mangle]
pub fn window_unlink(mut win: NullableDrc<window>, mut sess: NullableDrc<session>) {
    let mut l: NullableDrc<link> = NullableDrc::null();
    l = win.links.tqh_first.get();
    while !l.is_null() {
        if NullableDrc::ptr_eq(l.session.get(), sess) {
            break;
        }
        l = l.window_entry.tqe_next.get();
    }
    if !l.is_null() {
        link_delete(l, 1 as libc::c_int | 2 as libc::c_int);
    }
}
#[no_mangle]
pub fn link_delete(mut l: NullableDrc<link>, mut cleanup_flags: libc::c_int) {
    println!(
        "unlink session {}, window {}",
        l.session.get().id.get(),
        l.window.get().id.get(),
    );
    if cleanup_flags & 1 as libc::c_int != 0 {
        if !(l.session_entry.tqe_next.get()).is_null() {
            l.session_entry.tqe_next.get()
                .session_entry
                .tqe_prev.set(l.session_entry.tqe_prev.get());
        } else {
            l.session.get().links.tqh_last.set(l.session_entry.tqe_prev.get());
        }
        (*l.session_entry.tqe_prev.get()).set(l.session_entry.tqe_next.get());
        if (l.session.get().links.tqh_first.get()).is_null() {
            session_delete(l.session.get());
        }
    }
    if cleanup_flags & 2 as libc::c_int != 0 {
        if !(l.window_entry.tqe_next.get()).is_null() {
            l.window_entry.tqe_next.get()
                .window_entry
                .tqe_prev.set(l.window_entry.tqe_prev.get());
        } else {
            l.window.get().links.tqh_last.set(l.window_entry.tqe_prev.get());
        }
        (*l.window_entry.tqe_prev.get()).set(l.window_entry.tqe_next.get());
        if (l.window.get().links.tqh_first.get()).is_null() {
            window_delete(l.window.get());
        }
    }
    l.break_cycles();
}
fn main_0() -> libc::c_int {
    let mut sess1: NullableDrc<session> = session_new();
    let mut sess2: NullableDrc<session> = session_new();
    println!("\ndelete a window explicitly, which removes it from its sessions");
    let mut win3: NullableDrc<window> = window_new(sess1);
    window_link(win3, sess2);
    session_render(sess1);
    session_render(sess2);
    window_delete(win3);
    session_render(sess1);
    session_render(sess2);
    println!("\ndelete a window implicitly by removing it from all sessions");
    let mut win4: NullableDrc<window> = window_new(sess1);
    window_unlink(win4, sess1);
    session_render(sess1);
    session_render(sess2);
    println!("\ndelete a session implicitly by removing all of its windows");
    let mut sess3: NullableDrc<session> = session_new();
    let mut win5: NullableDrc<window> = sess3.links.tqh_first.get().window.get();
    window_unlink(win5, sess3);
    println!("\ndelete a session implicitly by deleting all of its windows");
    let mut sess4: NullableDrc<session> = session_new();
    let mut win6: NullableDrc<window> = sess4.links.tqh_first.get().window.get();
    window_delete(win6);
    println!("\ndelete sessions, which removes and deletes all of their windows");
    let mut win7: NullableDrc<window> = window_new(sess1);
    window_link(win7, sess2);
    session_render(sess1);
    session_render(sess2);
    session_delete(sess2);
    session_delete(sess1);
    0
}
pub fn main() {
    ::std::process::exit(main_0() as i32)
}
